cpp/include/raft/sparse/csr.hpp

```diff
@@ -178,5 +178,141 @@ void weak_cc(Index_* labels,
     labels, row_ind, row_ind_ptr, nnz, N, 0, N, stream, [](Index_) { return true; });
 }
 
+/**
+ * @brief CSR representation of a MFG.
+ *
+ * @tparam IdxT node index type
+ *
+ * Same as `mfg_ellpack` except the neighbors are in CSR format.
+ *
+ * @note Valid IDs: [0, 2^(sizeof(IdxT) * 8 - 1) - 2] (see `INVALID_ID`)
+ * @note This object does NOT own any of the underlying pointers and thus is
+ *       left to the calling code for their lifetime management.
+ * @note This is always an in-graph (`offsets` over the destination nodes).
+ *       See `fg_csr` for further explanation on this.
+ */
+template <typename IdxT>
+struct mfg_csr {
+  // handle sub-classes correctly
+  mfg_csr()          = default;
+  virtual ~mfg_csr() = default;
+
+  /**
+   * number of output nodes of this MFG. See `mfg_ellpack`.
+   * Note that input/output nodes are defined as in the regular MFG
+   */
+  IdxT n_out_nodes{0};
+  /** sample size: See `mfg_ellpack` */
+  IdxT sample_size{0};
+  /**
+   * number of input nodes of this MFG. See `mfg_ellpack`
+   * Note that input/output nodes are defined as in the regular MFG
+   */
+  IdxT* n_in_nodes{nullptr};
+  /**
+   * total number of edges in this graph (length of `indices` array)
+   * [single element, pinned memory]
+   */
+  IdxT* n_indices{nullptr};
+  /**
+   * Output nodes of this MFG. [len = `n_out_nodes`]
+   * For each output node, contains the index w.r.t. input nodes/features.
+   */
+  IdxT* out_nodes{nullptr};
+  /**
+   * Input nodes of this MFG. [len = `*n_in_nodes`]
+   * For each input node, contains the index w.r.t. input nodes/features of
+   * the previous layer or the global graph if no previous layer exists.
+   */
+  IdxT* in_nodes{nullptr};
+  /**
+   * monotonically increasing array with each location pointing to the start
+   * offset of the neighborhood of that node in the `indices` array. It is of
+   * length `n_out_nodes + 1`.
+   */
+  IdxT* offsets{nullptr};
+  /**
+   * contains neighbor indices of every node belonging to this object. It is of
+   * length `n_indices`. For each node, the number of neighbors is
+   * `<= sample_size`
+   */
+  IdxT* indices{nullptr};
```

Comment:
I am thinking about replacing this offsets + indices part with https://github.com/rapidsai/cugraph/blob/branch-22.06/cpp/include/cugraph/edge_partition_device_view.cuh#L36 from cugraph.

I think this class can be implemented with `edge_partition_device_view_base_t` + additional meta data (and cugraph's `edge_partition_device_view_t` for SG & MNMG are also implemented as `edge_partition_device_view_base_t` + additional meta data). We may move `edge_partition_device_view_base_t` to raft (with some renaming to match raft's conventinoo) and use this as the base class for adjacency list.

Reply:
Also want to mention that we have `edgelist_t` (https://github.com/rapidsai/cugraph/blob/branch-22.06/cpp/include/cugraph/graph.hpp#L33). I think this can also be moved to raft. Then we may have very basic adjacency list/edge list data structures to be extended elsewhere.

Reply:
I agree w/ @seunghwak here, I think these classes could be moved to RAFT (renamed accordingly). We have a COO (owning) class in RAFT currently which is being used by several primitives but I'd rather move the view classes from cugraph and use them throughout RAFT's sparse primitives.